cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -121,6 +136,60 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
     }
   }
 
+  // TODO CDAP-20343, Implement RepositoryManager#getFileHash()
+  public String getFileHash(String relativeFilePath, String commitID) throws IOException, NotFoundException {
+    return "file hash placeholder";
+  }
+
+  /**
+   * Commits and pushes all changes under the repository base path
+   *
+   * @param commitMeta Details for the commit including author, committer and commit message
+   * @throws UnexpectedRepositoryChangesException when there are changes outside the repository base path
+   * @throws GitAPIException                      when the underlying git commands fail
+   * @throws NoChangesToPushException             when there's no file changes for the commit
+   * @return the pushed commit ID
+   */
+  public String commitAndPush(CommitMeta commitMeta, List<String> filesChanged)
+    throws UnexpectedRepositoryChangesException, NoChangesToPushException, GitAPIException {
+    validateInitialized();
+
+    // if the status is clean skip
+    Status preStageStatus = git.status().call();
+    if (preStageStatus.isClean()) {
+      throw new NoChangesToPushException("No changes have been maid for the applications to push.");
+    }
+
+    for (String path : filesChanged) {
+      git.add().addFilepattern(path).call();
```

Comment:
The path that we get from the parameters may not be relative to the repository root. An optional path prefix can be configured in the repository config that is part of the reposBasePath.
Git expects the path to be relative to the repository root. In [this open PR](https://github.com/cdapio/cdap/pull/14917/files#diff-a3977abede2780220a49a40a806e42e944d4239be7e4c7cc302194260cf1c819), I've resolved the absoulute path, then made it relative to the respiratory root.

Reply:
@arjan-bal Actually in this PR, we made sure the passed in list is relative path to root, but just in String format. There's function `getAndValidateApplicationConfigRelativePath` that does it

Reply:
I see, so this is getting the path relative to `repositoryManager.getRepositoryRoot()` and not `repositoryManager.getBasePath()`. We should change to doc string so that it doesn't say bas path.

Reply:
right, will need to update that as well